File: ios/Sources/MultandoSDK/Services/VehicleTypeService.swift
import Foundation

/// Fetches and caches vehicle types.
public final class VehicleTypeService: @unchecked Sendable {

    private let httpClient: HTTPClient
    private var cache: [VehicleTypeResponse]?
    private var cacheTimestamp: Date?
    private let cacheTTL: TimeInterval = 300 // 5 minutes
    private let lock = NSLock()

    init(httpClient: HTTPClient) {
        self.httpClient = httpClient
    }

    /// List all vehicle types. Results are cached for 5 minutes.
    public func list(forceRefresh: Bool = false) async throws -> [VehicleTypeResponse] {
        if !forceRefresh, let cached = cachedValue() {
            return cached
        }

        let types: [VehicleTypeResponse] = try await httpClient.request(
            method: "GET",
            path: "/api/v1/vehicle-types"
        )

        lock.lock()
        cache = types
        cacheTimestamp = Date()
        lock.unlock()

        return types
    }

    /// Fetch a single vehicle type by ID.
    public func get(id: String) async throws -> VehicleTypeResponse {
        try await httpClient.request(
            method: "GET",
            path: "/api/v1/vehicle-types/\(id)"
        )
    }

    /// Clear the in-memory cache.
    public func clearCache() {
        lock.lock()
        cache = nil
        cacheTimestamp = nil
        lock.unlock()
    }

    private func cachedValue() -> [VehicleTypeResponse]? {
        lock.lock()
        defer { lock.unlock() }
        guard let cache, let ts = cacheTimestamp, Date().timeIntervalSince(ts) < cacheTTL else {
            return nil
        }
        return cache
    }
}

File: android/multando-sdk/src/main/kotlin/com/multando/sdk/services/VehicleTypeService.kt
package com.multando.sdk.services

import com.multando.sdk.core.HttpClient
import com.multando.sdk.models.VehicleTypeResponse

/**
 * Fetches and caches vehicle types.
 */
class VehicleTypeService internal constructor(
    private val httpClient: HttpClient
) {
    private var cache: List<VehicleTypeResponse>? = null
    private var cacheTimestamp: Long = 0
    private val cacheTtlMs: Long = 300_000 // 5 minutes

    /** List all vehicle types. Results are cached for 5 minutes. */
    suspend fun list(forceRefresh: Boolean = false): List<VehicleTypeResponse> {
        if (!forceRefresh) {
            cache?.let {
                if (System.currentTimeMillis() - cacheTimestamp < cacheTtlMs) return it
            }
        }

        val types = httpClient.request<List<VehicleTypeResponse>>(
            method = "GET",
            path = "/api/v1/vehicle-types"
        )

        synchronized(this) {
            cache = types
            cacheTimestamp = System.currentTimeMillis()
        }

        return types
    }

    /** Fetch a single vehicle type by ID. */
    suspend fun get(id: String): VehicleTypeResponse =
        httpClient.request(method = "GET", path = "/api/v1/vehicle-types/$id")

    /** Clear the in-memory cache. */
    fun clearCache() {
        synchronized(this) {
            cache = null
            cacheTimestamp = 0
        }
    }
}
